src/Psl/Encoding/Base64/url_decode.php

```diff
@@ -0,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Encoding\Base64;
+
+use Psl\Encoding\Exception;
+
+/**
+ * Decode a base64url-encoded string into raw binary.
+ *
+ * Base64url character set:
+ *  [A-Z]      [a-z]      [0-9]      -     _
+ *  0x41-0x5a, 0x61-0x7a, 0x30-0x39, 0x2d, 0x5f
+ *
+ * @pure
+ *
+ * @throws Exception\RangeException If the encoded string contains characters outside
+ *                                  the base64url characters range.
+ * @throws Exception\IncorrectPaddingException If the encoded string has an incorrect padding.
+ */
+function url_decode(string $base64url): string
+{
+    /** @var string */
+    return decode(strtr($base64url, '-_', '+/'));
```

Comment:
Should we be dealing with padding?

According to [RFC 4648 - section 5](https://datatracker.ietf.org/doc/html/rfc4648#section-5), the `=` symbol can be used in URLs as well.

Since PHP's base64 decode [deals with padding internally](https://github.com/php/php-src/blob/79d024ac0e2a07b93362a5a2fd4ef36f61fca8fb/ext/standard/base64.c#L1243-L1261), padding could be removed as well.

> The pad character "=" is typically percent-encoded when used in an
   URI [[9](https://datatracker.ietf.org/doc/html/rfc4648#ref-9)], but if the data length is known implicitly, this can be
   avoided by skipping the padding; see [section 3.2](https://datatracker.ietf.org/doc/html/rfc4648#section-3.2).

Reply:
I think `bool $padding = false` as an optional argument is the right call.

wdyt @veewee ?

Reply:
Just to be sure, when `false`, we remove padding in encode and add it again in decode, that's it?

Reply:
I commented it on the wrong function:
Whilst encoding the base64 it would rtrim the padding when padding is false.
PHP's base64 decode deals with padding - so you don't really have to take care from it there:

https://3v4l.org/nZZrv

Reply:
Is the enum really needed? With only 2 constants you can replace it with just a boolean:
```php
url_encode(string $binary, bool $explicitPadding = false): string
url_decode(string $base64url, bool $explicitPadding = false): string
```

Reply:
There won't be a 3rth option, so a boolean might make sense as well..
The `explicit` padding should be the default to keep the function BC compatible though.

Reply:
So in commit 4ed0d0e, I've added an argument for padding.

For encoding the argument is name `padding` to tell if we add padding or not. For decoding it's `explicitPadding` to tell if we check padding or not.
I've set these arguments to `true` for base64 and to `false` for base64url.

Reply:
🤔 Just wondering 2 things: 

* does it make sense to set these settings differently for regular vs url encoding? That would be rather inconsistent no?
* does it make sense to have 2 different names for the same thing in encode vs decode? Explicit padding works there as well maybe?

Reply:
- In my opinion, we don't have the same use between regular and url encoding. For regular encoding we want to keep padding as it can be needed (and rfc don't tell that we can remove it). However, for url encoding, as we pass the result in an url, to avoir percent encoding of = characters, it will be better to remove it by default.
- I see the things like that :
    - for encoding, we keep or not the padding, so the parameter tell if we want it or not.
    - for decoding, we have padding or not, so the parameter tell if we check it or not

Reply:
this conversation made me realize that i already implemented this somewhere else.

see:

https://github.com/nuxed/crypto/blob/develop/src/Nuxed/Crypto/Base64/encode.hack
https://github.com/nuxed/crypto/blob/develop/src/Nuxed/Crypto/Base64/decode.hack
https://github.com/nuxed/crypto/blob/develop/src/Nuxed/Crypto/Base64/UrlSafe/encode.hack
https://github.com/nuxed/crypto/blob/develop/src/Nuxed/Crypto/Base64/UrlSafe/decode.hack

and since we are adding more variants, it would make more sense to also add 

https://github.com/nuxed/crypto/tree/develop/src/Nuxed/Crypto/Base64/DotSlash/encode.hack
https://github.com/nuxed/crypto/tree/develop/src/Nuxed/Crypto/Base64/DotSlash/decode.hack
https://github.com/nuxed/crypto/tree/develop/src/Nuxed/Crypto/Base64/DotSlash/Ordered/encode.hack
https://github.com/nuxed/crypto/tree/develop/src/Nuxed/Crypto/Base64/DotSlash/Ordered/decode.hack

however, this would mean we end up with too many functions, what if we do:


```php
namespace Psl\Encoding\Base64;

enum Variant {
  case Default;
  case UrlSafe;
  case DotSlash;
  case DotSlashOrdered;
}

function encode(string $data, Variant $variant = Variant::Default, bool $pad = false): string { ... }
function decode(string $data, Variant $variant = Variant::Default, bool $pad = false): string { ... }
```

wdyt?

Reply:
Interesting! I don't know dot slash encoding neither hack language. I think I see how to simple dot slash (as for url safe just replace +/ by ./), but for dot slash ordered it seems more complicated to me. The solution I see is implement all base64 encoding as you have done for Nuxed/crypto. For that we can take inspiration from https://github.com/paragonie/constant_time_encoding

Reply:
> And same for decoding. It's a solution without use of class. 

Using internal classes is okay, multiple components do have classes in their internal namespace.

Reply:
:+1: Ok, i will implement that this evening (and maybe push the commit as well this evening if all goes well)